### PR TITLE
Hedgehog testing of pioneers exercise 1

### PR DESCRIPTION
**Pull request by Jimbo4350**
Tests the building of payment and stake addresses as described in [pioneers exercise 1](https://github.com/input-output-hk/cardano-tutorials/blob/master/pioneers-testnet/pioneers-exercise-1.md). The test failures leverage `optparse-applicative` errors.

### CODE REVIEW

File: cardano-cli/test/Test/Scenarios/OptParse.hs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import qualified Hedgehog as H
+import           Hedgehog.Internal.Property (failWith)
+
+-- | Purely evalutate the cardano-cli parser.
+-- e.g evalCardanoCLIParser ["shelley"] would be equivalent to cabal exec cardano-cli shelley
+-- without running underlying IO.
+evalCardanoCLIParser :: [String] -> Opt.ParserResult ClientCommand
+evalCardanoCLIParser args = Opt.execParserPure pref opts args
+
+-- | This takes a `ParserResult`, which is pure, and executes it.
```
Comment (intricate) [inline cardano-cli/test/Test/Scenarios/OptParse.hs:32]:
```suggestion
-- | This takes a 'ParserResult', which is pure, and executes it.
```

File: cardano-cli/test/Test/Scenarios/ShelleyAddressBuild.hs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    -- Build shelley address
+    execCardanoCLIParser
+      []
+      "Failure: prop_shelleyAddressBuild.build_address"
+        $ evalCardanoCLIParser [ "shelley", "address", "build"
+                               , "--payment-verification-key-file", addrVerKey
+                               ]
+
+    liftIO $ fileCleanup [addrVerKey, addrSignKey]
```
Comment (intricate) [inline cardano-cli/test/Test/Scenarios/ShelleyAddressBuild.hs:52]: Should we still attempt a file clean up in the failure case?

n.b. I see that `execCardanoCLIParser` will clean up in the failure case, but this is in reference to the `H.assert`s above.

File: cardano-cli/test/Test/Scenarios/ShelleyAddressBuild.hs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+-- | This property checks the scenario where we would like to
+-- build a shelley address.
+-- 1. We generate a key pair
+-- 2. We check for the existence of the key pair
+-- 3. We use the generated verification key to build a shelley address.
+prop_shelleyAddressBuild :: Property
+prop_shelleyAddressBuild =
+  H.withTests 1 . H.property $ do
```
Comment (intricate) [inline cardano-cli/test/Test/Scenarios/ShelleyAddressBuild.hs:23]: Why don't we run this more than once? Seems like it'd work fine.

File: cardano-cli/test/cardano-cli-test.hs [hunk 4]
```diff
@@ -22,8 +22,10 @@ import           System.Process (rawSystem)
 import           System.IO (BufferMode (..))
 import qualified System.IO as IO
 
+
```
Comment (intricate) [inline cardano-cli/test/cardano-cli-test.hs:25]: A bit of `diff` noise in this file :smile: 

```suggestion
```

File: cardano-cli/test/cardano-cli-test.hs [hunk 5]
```diff
@@ -22,8 +22,10 @@ import           System.Process (rawSystem)
...
 main :: IO ()
 main = do
+
```
Comment (intricate) [inline cardano-cli/test/cardano-cli-test.hs:28]:
```suggestion
```

File: cardano-cli/cardano-cli.cabal [hunk 6]
```diff
@@ -162,3 +161,34 @@ test-suite cardano-cli-test
...
                         -Wcompat
                         -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
+
+test-suite cardano-cli-scenarios
+  hs-source-dirs:       test
+  main-is:              cardano-cli-scenarios.hs
+  type:                 exitcode-stdio-1.0
+
+  if os(windows)
+    buildable: False
```
Comment (intricate) [inline cardano-cli/cardano-cli.cabal:171]: Is it true that this isn't buildable on Windows? I don't think we're doing anything *nix specific here, right?

File: cardano-cli/test/Test/Pioneers/Exercise1/BuildShelleyPaymentAddress.hs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+
+-- | This property checks the scenario where we would like to
+-- build a shelley address.
+-- 1. We generate a key pair
+-- 2. We check for the existence of the key pair
+-- 3. We use the generated verification key to build a shelley address.
+prop_buildShelleyPaymentAddress :: Property
+prop_buildShelleyPaymentAddress =
+  H.withTests 1 . H.property $ do
```
Comment (erikd) [inline cardano-cli/test/Test/Pioneers/Exercise1/BuildShelleyPaymentAddress.hs:22]: Apart from `H.withTests` and `H.property` what are we using from HH?

Could this not just be a function in `IO Bool` or similar? Ie returns `True` on success.

Comment (Jimbo4350) [reply cardano-cli/test/Test/Pioneers/Exercise1/BuildShelleyPaymentAddress.hs:22]: I was planning to use `Group` to group the exercises as we go along and `hedgehog` helps with the error rendering. If you are opposed to using `hedgehog` for these I can remove its usage.

Comment (erikd) [reply cardano-cli/test/Test/Pioneers/Exercise1/BuildShelleyPaymentAddress.hs:22]: I was not so much opposed to HH as wondering why. Nice error rendering actually makes sense.

So it you want it for the error rendering you should pull out a function like:

```
propertyOnce =  H.withTests 1 . H.property
```
and then use that everywhere.

**Comment by Jimbo4350** [comment]
bors merge

**Pull request merged**